core/shared/src/main/scala/zio/ZRef.scala

```diff
@@ -320,33 +340,68 @@ object ZRef extends Serializable {
         type S = self.S
         def getEither(s: S): Either[ED, D] =
           self.getEither(s).fold(e => Left(eb(e)), bc)
-        def setEither(c: C): Either[EC, S] =
-          ca(c).flatMap(a => self.setEither(a).fold(e => Left(ea(e)), Right(_)))
+        def setEither(c: C)(s: S): Either[EC, S] =
+          ca(c).flatMap(a => self.setEither(a)(s).fold(e => Left(ea(e)), Right(_)))
         val value: Atomic[S] =
           self.value
       }
 
-    final def fold0[EC, ED, C, D](
+    final def foldS[EC >: EB, ED, C, D](
       ea: EA => EC,
-      eb: EB => Either[ED, D],
-      ca: C => Either[EC, A],
+      eb: EB => ED,
+      ca: C => B => Either[EC, A],
       bc: B => Either[ED, D]
     ): ZRef[EC, ED, C, D] =
       new Derived[EC, ED, C, D] {
         type S = self.S
         def getEither(s: S): Either[ED, D] =
-          self.getEither(s).fold(eb, bc)
-        def setEither(c: C): Either[EC, S] =
-          ca(c).flatMap(a => self.setEither(a).fold(e => Left(ea(e)), Right(_)))
+          self.getEither(s).fold(e => Left(eb(e)), bc)
+        def setEither(c: C)(s: S): Either[EC, S] =
+          self.getEither(s).flatMap(ca(c)).flatMap(a => self.setEither(a)(s).fold(e => Left(ea(e)), Right(_)))
         val value: Atomic[S] =
           self.value
       }
 
     final def set(a: A): IO[EA, Unit] =
-      setEither(a).fold(ZIO.fail(_), value.set)
+      value.modify { s =>
+        setEither(a)(s) match {
+          case Left(e)  => (Left(e), s)
+          case Right(s) => (Right(()), s)
+        }
+      }.absolve
 
     final def setAsync(a: A): IO[EA, Unit] =
-      setEither(a).fold(ZIO.fail(_), value.setAsync)
+      value.modify { s =>
+        setEither(a)(s) match {
+          case Left(e)  => (Left(e), s)
+          case Right(s) => (Right(()), s)
+        }
+      }.absolve
+  }
```

Comment:
This was the main reason to add another type for `foldS` result in my PR.
These implementations are clearly more complex than older one, which can be significant in the 
And `setAsync` loses ability to delay modification.
So I felt like it's good to keep a somewhat performant version for simple mappings.

Reply:
Yes could definitely add this.

Reply:
Updated to add subtype to optimize `set` and `setAsync` for derived `ZRef` instances where possible.